File: learnLeetCode/Day26/Part4/portal_maze.cpp
#include <bits/stdc++.h>
using namespace std;

int min_steps(const vector<string>& maze){
    int n = (int)maze.size();
    int m = n? (int)maze[0].size(): 0;
    if(n==0 || m==0) return -1;
    unordered_map<char, vector<pair<int,int>>> portals;
    pair<int,int> S{-1,-1}, T{-1,-1};
    for(int i=0;i<n;++i){
        for(int j=0;j<m;++j){
            char c = maze[i][j];
            if(c=='S') S={i,j};
            else if(c=='T') T={i,j};
            else if('A'<=c && c<='Z') portals[c].push_back({i,j});
        }
    }
    if(S.first==-1 || T.first==-1) return -1;
    auto teleport = [&](int i,int j){
        char c = maze[i][j];
        if(!('A'<=c && c<='Z')) return pair{i,j};
        auto it = portals.find(c);
        if(it==portals.end()) return pair{i,j};
        auto &lst = it->second;
        if(lst.size()==2){
            if(lst[0]==pair{i,j}) return lst[1];
            else return lst[0];
        }
        return pair{i,j};
    };

    queue<pair<int,int>> q; q.push(S);
    vector dist(n, vector<int>(m, -1)); dist[S.first][S.second]=0;
    int dirs[4][2]={{-1,0},{1,0},{0,-1},{0,1}};
    while(!q.empty()){
        auto [x,y] = q.front(); q.pop();
        if(x==T.first && y==T.second) return dist[x][y];
        for(auto &d: dirs){
            int nx=x+d[0], ny=y+d[1];
            if(nx<0||nx>=n||ny<0||ny>=m) continue;
            if(maze[nx][ny]=='#') continue;
            pair<int,int> nxt;
            if(maze[nx][ny]=='.' || maze[nx][ny]=='S' || maze[nx][ny]=='T') nxt={nx,ny};
            else nxt = teleport(nx,ny);
            if(dist[nxt.first][nxt.second]==-1){
                dist[nxt.first][nxt.second]=dist[x][y]+1;
                q.push(nxt);
            }
        }
    }
    return -1;
}

int main(){
    vector<string> maze = {
        "S.A#",
        ".#A.",
        ".#..",
        "...T"
    };
    cout << min_steps(maze) << "\n"; // Expected 7
    return 0;
}

File: learnLeetCode_remote/Day26/Part4/portal_maze.py
from collections import deque
from typing import List, Tuple

def min_steps(maze: List[str]) -> int:
    n, m = len(maze), len(maze[0]) if maze else 0
    if n == 0 or m == 0:
        return -1
    portals = {}
    start = end = None
    for i in range(n):
        for j in range(m):
            c = maze[i][j]
            if c == 'S':
                start = (i, j)
            elif c == 'T':
                end = (i, j)
            elif 'A' <= c <= 'Z':
                portals.setdefault(c, []).append((i, j))
    if start is None or end is None:
        return -1

    def teleport(i: int, j: int) -> Tuple[int,int]:
        c = maze[i][j]
        if not ('A' <= c <= 'Z'): return (i,j)
        lst = portals.get(c, [])
        if len(lst) == 2:
            a, b = lst
            return b if (i,j) == a else a
        # single or more than 2: stay in place (robustness)
        return (i,j)

    q = deque([start])
    dist = {start: 0}
    DIRS = [(-1,0),(1,0),(0,-1),(0,1)]
    while q:
        x, y = q.popleft()
        if (x, y) == end:
            return dist[(x,y)]
        for dx, dy in DIRS:
            nx, ny = x + dx, y + dy
            if 0 <= nx < n and 0 <= ny < m and maze[nx][ny] != '#':
                if maze[nx][ny] == '.' or maze[nx][ny] in 'ST':
                    nxt = (nx, ny)
                else:
                    # letter: step into it then teleport to its pair
                    nxt = teleport(nx, ny)
                if nxt not in dist:
                    dist[nxt] = dist[(x,y)] + 1
                    q.append(nxt)
    return -1

if __name__ == "__main__":
    maze = [
        "S.A#",
        ".#A.",
        ".#..",
        "...T"
    ]
    print(min_steps(maze))  # Expected 7
